Ddpb 2724 bypass case used exception

slizzio opened the pull request:
## Purpose
To allow a lay deputy to register for a client when that client already exists under a professional deputyship.

Fixes [DDPB-2724](https://opgtransform.atlassian.net/browse/DDPB-2724)

## Approach
After much deliberation I have made the client soft deletable. This is to avoid removing the client completely and any reports that might be associated. So the action of discharging a client is to make them soft deleted. This prevents the client appearing anywhere in client. However admin area may sttill want to access the client (and reports) from the old professional deputyship so for admin pages I have disabled the soft delete filtter and re-enabled it once the query has been run. This allows the client to appear twice in the event that a lay deputy has re-registered against the case number. To clearly identify which is which I have applied a discharged date where required on the admin pages.

## Learning
_Any tips and tricks, blog posts or tools which helped you. Plus anything notable you've discovered about DigiDeps_

## Checklist
- [x] I have performed a self-review of my own code
- [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
- [x] I have added tests to prove my work, and they follow our [best practices](https://github.com/ministryofjustice/opg-digi-deps-client/wiki/Testing-best-practices)
- [x] I have successfully built my branch to a feature environment
- [x] New and existing unit tests pass locally with my changes (`docker-compose run --rm test sh scripts/clienttest.sh`)
- [x] There are no new frontend linting errors (`docker-compose run --rm npm run lint`)
- [ ] There are no NPM security issues (`docker-compose run --rm npm audit`)
- [x] There are no Composer security issues (`docker-compose run frontend php app/console security:check`)
- [ ] The product team have tested these changes

### For each page changed
- [ ] I have run an in-browser accessibility test (e.g. WAVE, Lighthouse)
- [ ] There are no deprecated CSS classes noted in the profiler
- [ ] Translations are used and the profiler doesn't identify any missing

slizzio committed:
DDPB-2724 Fixed bug when identifying previous existing report

slizzio committed:
DDPB-2724 Fixed bug in report Ids during self reg flow

slizzio committed:
DDPB-2724 Add Soft deletable trait to Client

slizzio committed:
DDPB-2724 revise report submissions table to reflect discharged deputies

slizzio committed:
DDPB-2724 added activeAuditt to return virtual properies of from and to date for a client

slizzio committed:
DDPB-2724 revised active dates for client, added content to reflect period of deputyship and discharged date

slizzio committed:
DDPB-2724 Added behat test for discharged client

slizzio committed:
DDPB-2724 Added registration steps against new client with discharged client case number

slizzio committed:
DDPB-2724 added more checks for deleted client

slizzio committed:
DDPB-2724 Fixed tests

gregtyler commented on src/AppBundle/Entity/Traits/ActiveAudit.php:
```diff
@@ -0,0 +1,67 @@
+//    {
+//        $this->activeFrom = $activeFrom;
+//        return $this;
+//    }
```
Should remove this code rather than commenting it out

slizzio replied:
Done

gregtyler commented on src/AppBundle/Entity/Traits/ActiveAudit.php:
```diff
@@ -0,0 +1,67 @@
+//    {
+//        $this->activeTo = $activeTo;
+//        return $this;
+//    }
```
Ditto RE comments

gregtyler commented on src/AppBundle/Resources/views/Admin/Client/Client/details.html.twig:
```diff
@@ -23,12 +23,14 @@
     {% endif %}
     <p>{{ 'courtOrderNumber' | trans({}, 'common') }}: <span class="bold">{{ client.caseNumber }}</span></p>
+    {% if client.deletedAt %}
+        <p class="flush--bottom behat-region-discharged-on">Discharged on: <span class="bold">{{ client.deletedAt | date("j M Y") }}</span></p>
```
"Discharged on" should be in a translation.

Also, this little definition unit should be [summary list](https://design-system.service.gov.uk/components/summary-list/) rather than a bunch of `<p>`s. I know it's not introduced by your changes, but I think it would be good to change that whilst we're here.

slizzio replied:
Done

gregtyler replied:
Nice one 💪

gregtyler commented on src/AppBundle/Resources/views/Admin/Client/Report/checklist.html.twig:
```diff
@@ -57,6 +57,10 @@
             </p>
+            {% if report.client.deletedAt %}
+                <p>{{ 'dischargedOn' | trans({}, 'common') }}:
+                    Discharged on: <span class="bold behat-region-discharged-on">{{ report.client.deletedAt | date("j M Y") }}</span></p>
```
Again, "Discharged on" should be a translation

gregtyler commented on tests/behat/bootstrap/DbTrait.php:
```diff
@@ -109,4 +109,16 @@ public function IDeleteTheAppSetting($setting)
+     */
+    public function iDischargeTheDeputiesFromCase($caseNumber)
+    {
+//        $query = "DELETE FROM deputy_case where client_id = (SELECT id FROM client WHERE case_number = '{$caseNumber}'";
```
Should remove this `DELETE FROM` comment to avoid confusion

slizzio replied:
removed

gregtyler commented on tests/behat/features/prof/03-report/09-deputy-discharged.feature:
```diff
@@ -0,0 +1,202 @@
+
+  Scenario: Case manager discharges professional deputy from client
+    Given I am logged in to admin as "casemanager@publicguardian.gov.uk" with password "Abcd1234"
+    And I discharge the deputies from case "01000010"
```
Can we add some earlier test steps to ensure that the discharged regions aren't shown _before_ the deputy is discharged?

Also, this test should propbably use more Whens and Thens.

slizzio replied:
Amended tests, look much better now. It also highlighted the discrepancy when two clients of same case number appear in the list. Ive amended behat regions these with "discharged-client-XXXX" to allow tests to click on the appropriate region.

slizzio committed:
DDPB-2724 removed comments

slizzio committed:
DDPB-2724 amended client list to reflect discharged deputy snd amended tests accordingly

gregtyler commented on src/AppBundle/Resources/views/Admin/Client/Client/details.html.twig:
```diff
@@ -24,11 +24,53 @@
+    {#<p>{{ 'courtOrderNumber' | trans({}, 'common') }}: <span class="bold">{{ client.caseNumber }}</span></p>#}
+    {#{% if client.deletedAt %}#}
+        {#<p class="flush--bottom behat-region-discharged-on">Discharged on: <span class="bold">{{ client.deletedAt | date("j M Y") }}</span></p>#}
+    {#{% endif %}#}
```
Should probably ditch this commented out code, not a blocker though

slizzio replied:
Revised to show a summary as per previous review. Should not have been commented

slizzio committed:
DDPB-2724 Moved new files into client folder

slizzio committed:
DDPB-2724 Removed old code

slizzio committed:
DDPB-2782 Disabled delete user pending bug fix (its deleting all attached clients if first client has no reports)

slizzio committed:
DDPB-2724 limited delete user to lay deps with 1 client and no reports

slizzio committed:
DDPB-2724 fixed client count to limit to 1

slizzio committed:
Merge branch 'master' of github.com:ministryofjustice/opg-digi-deps-client into DDPB-2724-bypass-case-used-exception

gregtyler committed:
Drop expected file size by ~15%

DDPB-2809 changed the way PDF files are generated, which has led to smaller file sizes. We need to reduce the size expectations in the tests to match this, since they now sometimes fall below previous levels.

gregtyler commented:
[Transferred to opg-digideps](https://github.com/ministryofjustice/opg-digideps/pull/1)
